Source/bmalloc/libpas/src/libpas/pas_report_crash.c

```diff
@@ -76,100 +87,91 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
-            return KERN_FAILURE;
-        if (!temp_entry)
+        if (!pas_enumerator_copy_remote(enumerator, &hash_map_entry, hash_map.table + i, sizeof(pas_ptr_hash_map_entry))) {
+            pas_enumerator_destroy(enumerator);
             return KERN_FAILURE;
-        hash_map_entry = *temp_entry;
+        }
 
         /* Skip entry if not there */
         if (hash_map_entry.key == (void*)UINTPTR_MAX)
             continue;
 
-        pas_pgm_storage* temp_metadata = NULL;
-        kr = reader(task, (vm_address_t)hash_map_entry.value, sizeof(pas_pgm_storage), (void**)&temp_metadata);
-        if (kr != KERN_SUCCESS)
-            return KERN_FAILURE;
-        if (!temp_metadata)
+        if (!pas_enumerator_copy_remote(enumerator, &pgm_metadata, hash_map_entry.value, sizeof(pas_pgm_storage))) {
+            pas_enumerator_destroy(enumerator);
             return KERN_FAILURE;
-        pgm_metadata = *temp_metadata;
+        }
 
-        pas_backtrace_metadata* resolved_alloc_backtrace = NULL;
+        pas_backtrace_metadata alloc_backtrace_data;
         if (pgm_metadata.alloc_backtrace) {
-            kr = reader(task, (vm_address_t)pgm_metadata.alloc_backtrace, sizeof(pas_backtrace_metadata), (void**)&resolved_alloc_backtrace);
-            if (kr != KERN_SUCCESS)
+            if (!pas_enumerator_copy_remote(enumerator, &alloc_backtrace_data, pgm_metadata.alloc_backtrace, sizeof(pas_backtrace_metadata))) {
+                pas_enumerator_destroy(enumerator);
                 return KERN_FAILURE;
-            if (!resolved_alloc_backtrace)
-                return KERN_FAILURE;
-            if (resolved_alloc_backtrace->frame_size < 0 || resolved_alloc_backtrace->frame_size > PGM_BACKTRACE_MAX_FRAMES)
+            }
+            if (alloc_backtrace_data.frame_size < 0 || alloc_backtrace_data.frame_size > PGM_BACKTRACE_MAX_FRAMES) {
+                pas_enumerator_destroy(enumerator);
                 return KERN_FAILURE;
+            }
         }
 
-        pas_backtrace_metadata* resolved_dealloc_backtrace = NULL;
+        pas_backtrace_metadata dealloc_backtrace_data;
         if (pgm_metadata.dealloc_backtrace) {
-            kr = reader(task, (vm_address_t)pgm_metadata.dealloc_backtrace, sizeof(pas_backtrace_metadata), (void**)&resolved_dealloc_backtrace);
-            if (kr != KERN_SUCCESS)
-                return KERN_FAILURE;
-            if (!resolved_dealloc_backtrace)
+            if (!pas_enumerator_copy_remote(enumerator, &dealloc_backtrace_data, pgm_metadata.dealloc_backtrace, sizeof(pas_backtrace_metadata))) {
+                pas_enumerator_destroy(enumerator);
                 return KERN_FAILURE;
-            if (resolved_dealloc_backtrace->frame_size < 0 || resolved_dealloc_backtrace->frame_size > PGM_BACKTRACE_MAX_FRAMES)
+            }
+            if (dealloc_backtrace_data.frame_size < 0 || dealloc_backtrace_data.frame_size > PGM_BACKTRACE_MAX_FRAMES) {
+                pas_enumerator_destroy(enumerator);
```

Comment:
I noticed quite a few instances of the below code repeated over again; not sure if we can simplify this at all.

```
pas_enumerator_destroy(enumerator);
return KERN_FAILURE;
```